src/wallet/rpcwallet.cpp

```diff
@@ -3748,6 +3749,340 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
+                {
+                    maxedOutUTXOsFlag = true;
+                } else {
+                    estimatedTxSize += increase;
+                    COutPoint utxo(out.tx->GetHash(), out.i);
+                    utxoInputs.emplace_back(utxo, nValue);
+                    mergedUTXOValue += nValue;
+                }
+            }
+
+            if (maxedOutUTXOsFlag) {
+                remainingUTXOValue += nValue;
+            }
+        }
+    }
+
+    if (useAny || useAnyNote || zaddrs.size() > 0) {
+        // Get available notes
+        std::vector<CNotePlaintextEntry> entries;
+        pwalletMain->GetFilteredNotes(entries, zaddrs);
+
+        // Find unspent notes and update estimated size
+        for (CNotePlaintextEntry& entry : entries) {
+            noteCounter++;
+            CAmount nValue = entry.plaintext.value;
+
+            if (!maxedOutNotesFlag) {
+                // If we haven't added any notes yet and the merge is to a
+                // z-address, we have already accounted for the first JoinSplit.
+                size_t increase = (noteInputs.empty() && !isToZaddr) || (noteInputs.size() % 2 == 0) ? JOINSPLIT_SIZE : 0;
+                if (estimatedTxSize + increase >= MAX_TX_SIZE ||
+                    (nNoteLimit > 0 && noteCounter > nNoteLimit))
+                {
+                    maxedOutNotesFlag = true;
+                } else {
+                    estimatedTxSize += increase;
+                    SpendingKey zkey;
+                    pwalletMain->GetSpendingKey(entry.address, zkey);
+                    noteInputs.emplace_back(entry.jsop, entry.plaintext.note(entry.address), nValue, zkey);
+                    mergedNoteValue += nValue;
+                }
+            }
+
+            if (maxedOutNotesFlag) {
+                remainingNoteValue += nValue;
+            }
+        }
+    }
+
+    size_t numUtxos = utxoInputs.size();
+    size_t numNotes = noteInputs.size();
+
+    if (numUtxos == 0 && numNotes == 0) {
+        throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "Could not find any funds to merge.");
+    }
+
+    // Sanity check: Don't do anything if:
+    // - We only have one from address
+    // - It's equal to toaddress
+    // - The address only contains a single UTXO or note
+    if (setAddress.size() == 1 && setAddress.count(destaddress) && (numUtxos + numNotes) == 1) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Destination address is also the only source address, and all its funds are already merged.");
+    }
+
+    CAmount mergedValue = mergedUTXOValue + mergedNoteValue;
+    if (mergedValue < nFee) {
+        throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS,
+            strprintf("Insufficient funds, have %s, which is less than miners fee %s",
+            FormatMoney(mergedValue), FormatMoney(nFee)));
+    }
+
+    // Check that the user specified fee is sane (if too high, it can result in error -25 absurd fee)
```

Comment:
z_mergetoaddress created a transaction with 8 joinsplits... when it finished, it triggered a mempool rejection due to absurd fee.

Command line:
```
./zcash-cli z_mergetoaddress '["zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i", "ztn8nEZgVxNGLHjC7xhCAdjBh4DC8fSWVND6GtGrhpSUAtoA6MbSuTeeo8eSCK9BcDvuU3nUuKWxGM6RGzfCSBoDK6Fdag8"]' ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE 3 3
{
  "remainingUTXOs": 0,
  "remainingTransparentValue": 0.00000000,
  "remainingNotes": 2,
  "remainingShieldedValue": 3.12480000,
  "mergingUTXOs": 0,
  "mergingTransparentValue": 0.00000000,
  "mergingNotes": 10,
  "mergingShieldedValue": 3045.74586597,
  "opid": "opid-20af324c-986d-42e3-b243-8db25df5b3d1"
}
```
z_getoperationstatus returned:
```
  {
    "id": "opid-20af324c-986d-42e3-b243-8db25df5b3d1",
    "status": "failed",
    "creation_time": 1519283425,
    "error": {
      "code": -25,
      "message": ""
    },
    "method": "z_mergetoaddress",
    "params": {
      "fromaddresses": [
        "zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i",
        "ztn8nEZgVxNGLHjC7xhCAdjBh4DC8fSWVND6GtGrhpSUAtoA6MbSuTeeo8eSCK9BcDvuU3nUuKWxGM6RGzfCSBoDK6Fdag8"
      ],
      "toaddress": "ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE",
      "fee": 3.00000000
    }
  }
```
and debug.log
```
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending 3045.74586597 to send 3042.74586597 with fee 3.00
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: transparent input: 0.00
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: private input: 3045.74586597
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: private output: 3042.74586597
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: fee: 3.00
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=da408a2512, vjoinsplit=2, ciphertext=0, amount=2560.99686597, height=662, confirmations=15)
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=fdee405f5e, vjoinsplit=0, ciphertext=0, amount=1.5624, height=670, confirmations=7)
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for change (amount=2562.55926597)
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 0 (vpub_old=0.00, vpub_new=0.00, in[0]=2560.99686597, in[1]=1.5624, out[0]=0.00, out[1]=2562.55926597)
2018-02-22 07:37:01 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=0, n=1, zaddr=ztqRjKVqZuyBS4dBxce1qLNmbRs2DaRFMtRFu7dfgcmLVVZeH7HhgciWWkmTE7CE65EXEyVut3YjYDG8nCP7FheKjsunHJS
2018-02-22 07:37:01 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=0, n=0, zaddr=zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i
2018-02-22 07:37:01 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending change (amount=2562.55926597)
2018-02-22 07:37:01 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=50d6136c79, vjoinsplit=0, ciphertext=1, amount=1.5624, height=667, confirmations=10)
2018-02-22 07:37:01 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for change (amount=2564.12166597)
2018-02-22 07:37:01 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 1 (vpub_old=0.00, vpub_new=0.00, in[0]=2562.55926597, in[1]=1.5624, out[0]=0.00, out[1]=2564.12166597)
2018-02-22 07:37:38 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=1, n=1, zaddr=ztntUYUSMPS7SxYr3TWkzsJdJWJeLkX1Rdgdz1hC66678otpLdHRpd8zFQjgoERd2CqS1wv25qbPLMFFdgjLgF5UiECBaCW
2018-02-22 07:37:38 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=1, n=0, zaddr=zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i
2018-02-22 07:37:38 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending change (amount=2564.12166597)
2018-02-22 07:37:38 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=65822fd89e, vjoinsplit=0, ciphertext=0, amount=1.5624, height=665, confirmations=12)
2018-02-22 07:37:38 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for change (amount=2565.68406597)
2018-02-22 07:37:38 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 2 (vpub_old=0.00, vpub_new=0.00, in[0]=2564.12166597, in[1]=1.5624, out[0]=0.00, out[1]=2565.68406597)
2018-02-22 07:38:18 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=2, n=1, zaddr=ztmpsYXY9TVQnuMDyChYcoxmKo2BPoZaDnTuWgVniY3Jz15zW7VF2tCtQJQubTDmUeTzQYBhCtsoWVnvFiNPv7dUFEzX2Wp
2018-02-22 07:38:18 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=2, n=0, zaddr=zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i
2018-02-22 07:38:18 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending change (amount=2565.68406597)
2018-02-22 07:38:18 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=9350300fe4, vjoinsplit=0, ciphertext=1, amount=237.8748, height=673, confirmations=4)
2018-02-22 07:38:18 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for change (amount=2803.55886597)
2018-02-22 07:38:18 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 3 (vpub_old=0.00, vpub_new=0.00, in[0]=2565.68406597, in[1]=237.8748, out[0]=0.00, out[1]=2803.55886597)
2018-02-22 07:38:57 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=3, n=1, zaddr=ztkMUzdEqdejnfvfWkCaMkQNGKkGrffuodLA7Y6XjopFKFr4uY67b1gVeypQTH6y75pdBHrFn5r8gatcoVwY2nsUV35jWuh
2018-02-22 07:38:57 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=3, n=0, zaddr=zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i
2018-02-22 07:38:57 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending change (amount=2803.55886597)
2018-02-22 07:38:57 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=1770d39e76, vjoinsplit=0, ciphertext=1, amount=1.5624, height=669, confirmations=8)
2018-02-22 07:38:57 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for change (amount=2805.12126597)
2018-02-22 07:38:57 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 4 (vpub_old=0.00, vpub_new=0.00, in[0]=2803.55886597, in[1]=1.5624, out[0]=0.00, out[1]=2805.12126597)
2018-02-22 07:39:34 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=4, n=0, zaddr=ztN5FRhAriHE9bmxfeVx2AuhFPQM9TK3fiLMDD6fRbF9cu1319Mm4szEFiHYraSpSy6XXw7bKWHSensZvYXTswXfFcwkLJM
2018-02-22 07:39:34 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=4, n=1, zaddr=zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i
2018-02-22 07:39:34 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending change (amount=2805.12126597)
2018-02-22 07:39:34 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=7e093d2378, vjoinsplit=0, ciphertext=1, amount=1.5624, height=666, confirmations=11)
2018-02-22 07:39:34 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for change (amount=2806.68366597)
2018-02-22 07:39:34 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 5 (vpub_old=0.00, vpub_new=0.00, in[0]=2805.12126597, in[1]=1.5624, out[0]=0.00, out[1]=2806.68366597)
2018-02-22 07:40:11 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=5, n=1, zaddr=ztP4FsgajHhwas86iLsd4vsBCBkuhzdAMQXZT3oMwc835UMRwdjMrsyY3KD3sbRnKgwpmRxPpbkK1uhsKht1TDAdEsCUS2j
2018-02-22 07:40:11 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=5, n=0, zaddr=zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i
2018-02-22 07:40:11 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending change (amount=2806.68366597)
2018-02-22 07:40:11 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=8d3a2e58dc, vjoinsplit=0, ciphertext=1, amount=1.5624, height=668, confirmations=9)
2018-02-22 07:40:11 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for change (amount=2808.24606597)
2018-02-22 07:40:11 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 6 (vpub_old=0.00, vpub_new=0.00, in[0]=2806.68366597, in[1]=1.5624, out[0]=0.00, out[1]=2808.24606597)
2018-02-22 07:40:49 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=6, n=1, zaddr=ztTkSnFnan48jUvre4qkABFGUSkzswXxcEkHbhsBaAEe87imo4yuhpHkQNd8hRnFESPGDm9cHxYgRbH3eKtUWS39KiRMhAn
2018-02-22 07:40:49 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=6, n=0, zaddr=zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i
2018-02-22 07:40:49 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending change (amount=2808.24606597)
2018-02-22 07:40:49 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=3d42639d4d, vjoinsplit=0, ciphertext=0, amount=235.9374, height=662, confirmations=15)
2018-02-22 07:40:49 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for change (amount=3044.18346597)
2018-02-22 07:40:49 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 7 (vpub_old=0.00, vpub_new=0.00, in[0]=2808.24606597, in[1]=235.9374, out[0]=0.00, out[1]=3044.18346597)
2018-02-22 07:41:26 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=7, n=1, zaddr=ztK5XQzh4ggA3vaaDeQoxXmLiGA8M2sXAXDonC72dtv7iP8zHkqdqEJ3ThDc6iCFA7pV3UWPGdgUDo4fCHtqT9S3FFuidb9
2018-02-22 07:41:26 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=7, n=0, zaddr=zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i
2018-02-22 07:41:26 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending change (amount=3044.18346597)
2018-02-22 07:41:26 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=62015dfca2, vjoinsplit=0, ciphertext=0, amount=1.5624, height=672, confirmations=5)
2018-02-22 07:41:26 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for target (amount=3042.74586597)
2018-02-22 07:41:26 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 8 (vpub_old=0.00, vpub_new=3.00, in[0]=3044.18346597, in[1]=1.5624, out[0]=0.00, out[1]=3042.74586597)
2018-02-22 07:42:04 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=8, n=1, zaddr=ztKqq3VXPFQnk3fUTGjq9YcnovGnHyNMLBJgPbZpV7En3xqEdzKiUZfVFVBmEoHnTeo4S7Kykt6nwTdvXyQj6JyZJbAc2ym
2018-02-22 07:42:04 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=8, n=0, zaddr=ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE
2018-02-22 07:42:04 ERROR: AcceptToMemoryPool: absurdly high fees 3c581d17a73c8ab675b0e128db3d4140f5d14c3258e9ac3f1a53f09823e01979, 300000000 > 16320000
2018-02-22 07:42:04 opid-20af324c-986d-42e3-b243-8db25df5b3d1: z_mergetoaddress finished (status=failed, error=)
```

Reply:
Not sure what to say there: you set a fee of 3 ZEC in the CLI, which was larger than what the mempool calculated as a maximum reasonable fee. `z_mergetoaddress` only checks beforehand that the fee is not greater than the value being merged, for which an absurd fee would of course be triggered, but I don't think it's necessary to duplicate the absurd fee logic everywhere else.

Reply:
Actually the fee of 3 was a typo :-)

There is a general point here where a user can't predict how big the transaction will be and what an appropriate fee is.   Also, in this call, merging 3045 ZEC for a fee of 3 ZEC, which is 0.1%, might be reasonable to some people, especially if they are sending to someone, rather than merging just for themselves. 

Here's the mempool check.

```
        if (fRejectAbsurdFee && nFees > ::minRelayTxFee.GetFee(nSize) * 10000)
            return error("AcceptToMemoryPool: absurdly high fees %s, %d > %d",
                         hash.ToString(),
                         nFees, ::minRelayTxFee.GetFee(nSize) * 10000);
```

It would be nice if we could add another check before spending a few minutes creating multiple joinsplits.  Since we know the mempool check is based on size, and we already compute `estimatedTxSize`, the following patch works and has been tested.

```
    if (nFee > netAmount) {
        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Fee %s is greater than the net amount to be shielded %s", FormatMoney(nFee), FormatMoney(netAmount)));
    }

    CAmount estimatedFee = minRelayTxFee.GetFee(estimatedTxSize);
    CAmount estimatedAbsurdLimit = estimatedFee * 10000; // Matches code in AcceptToMemoryPool
    if (nFee > estimatedAbsurdLimit) {
        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Fee %s is much bigger than the estimated fee %s", FormatMoney(nFee), FormatMoney(estimatedFee)));
    }
```

Here is some output with debug statements.

```
./zcash-cli z_mergetoaddress '["*"]' ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE 3
nfee = 300000000
netAmount = 305387036597
estimatedTxSize = 3804
estimatedFee = 380
estimatedAbsurd = 3800000
error code: -8
error message:
Fee 3.00 is much bigger than the estimated fee 0.0000038

./zcash-cli z_mergetoaddress '["*"]' ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE 0.04
nfee = 4000000
netAmount = 305683036597
estimatedTxSize = 3804
estimatedFee = 380
estimatedAbsurd = 3800000
error code: -8
error message:
Fee 0.04 is much bigger than the estimated fee 0.0000038

z_mergetoaddress '["*"]' ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE 0.038
nfee = 3800000
netAmount = 305683336597
estimatedTxSize = 3804
estimatedFee = 380
estimatedAbsurd = 3800000
{
  "remainingUTXOs": 0,
  "remainingTransparentValue": 0.00000000,
  "remainingNotes": 0,
  "remainingShieldedValue": 0.00000000,
  "mergingUTXOs": 0,
  "mergingTransparentValue": 0.00000000,
  "mergingNotes": 2,
  "mergingShieldedValue": 3056.87036597,
  "opid": "opid-e9d944ae-7ad0-48c2-bfdb-886ce393cd2d"
}
```

Reply:
Is this something we need to fix right now? Not to appeal to authority, but I copied this fee logic from your implementation of `z_shieldcoinbase` :stuck_out_tongue_winking_eye: I would prefer that the RPC is consistent, so if we make this kind of change, we should make it in a separate PR to all our `z_*` spending RPCs.

Reply:
ticket filed